console/v1/types_console_sample.go

```diff
@@ -0,0 +1,198 @@
+	// +optional
+	// +listType=set
+	// +kubebuilder:validation:MaxItems:=10
+	Tags []string `json:"tags"`
+
+	// The actual sample source, like an external git repository or container image.
+	// +kubebuilder:validation:Required
+	Source ConsoleSampleSource `json:"source"`
+}
+
+// ConsoleSampleSourceType is an enumeration of the supported sample types.
+// Unsupported samples types will be ignored in the web console.
+// +kubebuilder:validation:Enum:=GitImport;ContainerImport
+type ConsoleSampleSourceType string
+
+const (
+	// A sample that let the user import code from a git repository.
+	GitImport ConsoleSampleSourceType = "GitImport"
+	// A sample that let the user import a container image.
+	ContainerImport ConsoleSampleSourceType = "ContainerImport"
+)
+
+// ConsoleSampleSource is the actual sample definition and can hold different sample types.
+// Unsupported sample types will be ignored in the web console.
+// +union
+// +kubebuilder:validation:XValidation:rule="self.type == 'GitImport' ? has(self.gitImport) : !has(self.gitImport)",message="source.gitImport is required when source.type is GitImport, and forbidden otherwise"
+// +kubebuilder:validation:XValidation:rule="self.type == 'ContainerImport' ? has(self.containerImport) : !has(self.containerImport)",message="source.containerImport is required when source.type is ContainerImport, and forbidden otherwise"
+type ConsoleSampleSource struct {
+	// type of the sample, currently supported: "GitImport";"ContainerImport"
+	// +unionDiscriminator
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum:="GitImport";"ContainerImport"
+	Type ConsoleSampleSourceType `json:"type"`
+
+	// gitimport let the user import code from a git repository.
+	// +unionMember
+	// +optional
+	GitImport *ConsoleSampleGitImportSource `json:"gitImport,omitempty"`
+
+	// containerimport let the user import a container image.
+	// +unionMember
+	// +optional
+	ContainerImport *ConsoleSampleContainerImportSource `json:"containerImport,omitempty"`
+}
+
+// ConsoleSampleGitImportSource let the user import code from a public git repository.
+type ConsoleSampleGitImportSource struct {
+	// url of the Git repository. Only public HTTPS repositories on GitHub, GitLab
+	// and Bitbucket are currently supported:
+	//
+	//   - https://github.com/<org>/<repository>
+	//   - https://gitlab.com/<org>/<repository>
+	//   - https://bitbucket.org/<org>/<repository>
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=256
+	// +kubebuilder:validation:Pattern=`^https:\/\/(github.com|gitlab.com|bitbucket.org)\/[a-zA-Z0-9-]+\/[a-zA-Z0-9-]+(.git)?$`
+	URL string `json:"url"`
+	// revision to clone a specific branch, tag or commit SHA, etc.
+	// +optional
+	// +kubebuilder:validation:MaxLength=256
+	Revision string `json:"revision"`
+	// contextDir, a subfolder in the repo to build the component.
```

Comment:
```suggestion
	// contextDir is used to specify a directory within the repository to build the
	// component.
	// Must start with `/` and have a maximum length of 256 characters.
	// When omitted, the default value is to build from the root of the repository.
```
And drop the line below about starting with a slash